File: 46-permutations/46-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> sol = new ArrayList<>();
        int[] visited = new int[nums.length];
        dfs(nums, res, sol, visited);
        return res;
    }
    public void dfs(int[] nums, List<List<Integer>> res,  List<Integer> sol, int[] visited) {
        //base case
        if(sol.size() == nums.length) {
            res.add(new ArrayList<>(sol));
            return;
        }
        
        for(int i = 0; i < nums.length; i++) {
            if(visited[i]== 1) continue;
            sol.add(nums[i]);
            visited[i] = 1;
            dfs(nums, res, sol, visited);
            sol.remove(sol.size() - 1);
            visited[i] = 0;
        }
    }
}

File: 46-permutations/46-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        lst = []
        seen = set()
        def dfs():
            nonlocal res, lst, seen
            if len(lst) == len(nums):
                res.append(lst[:])
                return
            for i in range(len(nums)):
                if nums[i] in seen:
                    continue
                seen.add(nums[i])
                lst.append(nums[i])
                dfs()
                seen.remove(nums[i])
                lst.pop(-1)
        dfs()
        return res
                